### PR TITLE
PS kernel driver changes for KDS

### PR DESCRIPTION
**Pull request by jeffli-xilinx**
Add scu subdev
Use cu_index above MAX_CUS for PS kernels

### CODE REVIEW

**Review by maxzhen — APPROVED** [review]

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drv.c [hunk 1]
```diff
@@ -1275,12 +1278,14 @@ static int xocl_cma_mem_alloc_huge_page(struct xocl_dev *xdev, struct drm_xocl_a
 {
 	int ret = 0;
 	size_t page_sz = cma_info->total_size/cma_info->entry_num;
-	uint32_t i, j, num = MAX_SB_APERTURES;
+	uint32_t i, j, num = xocl_addr_translator_get_entries_num(xdev);
```
Comment (houlz0507) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drv.c:1281]: Why does this eliminate https://github.com/Xilinx/XRT/commit/54fc3976ba0df54657c1a2c721fe57b4336c8719 ?

Comment (jeffli-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drv.c]: Looks like I forgot to merge latest XRT master changes. 
I have updated the file with the correct changes.

**Review by houlz0507 — APPROVED** [review]

**Review by chienwei-lan — APPROVED** [review]

**Comment by dayeh-xilinx** [comment]
retest this please. Looks like an infra error.

File: src/runtime_src/core/common/drv/include/kds_core.h [hunk 2]
```diff
@@ -58,11 +58,13 @@ struct kds_ctx_info {
  * Let's see if we can unify them later.
  */
 struct kds_scu_mgmt {
+	struct xrt_cu		 *xcus[MAX_CUS];
```
Comment (mamin506) [inline src/runtime_src/core/common/drv/include/kds_core.h:61]: Maybe we need another struct for PS kernel. But right now, subdev/scu.c is not really talking to PS, so this is okay to me.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 3]
```diff
@@ -1580,6 +1580,64 @@ static int icap_create_subdev_cu(struct platform_device *pdev)
...
+	struct xrt_cu_info info;
+	int err = 0, i, j;
+	int inst = 0;
+
+	/* Let SCU controller know the dynamic resources */
+	for (i = 0; i < ps_kernel->pkn_count; ++i) {
+		scu_data = &ps_kernel->pkn_data[i];
+
+		for (j=0; j < scu_data->pkd_num_instances; ++j) {
+			struct xocl_subdev_info subdev_info = XOCL_DEVINFO_SCU;
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:1599]: XOCL_DEVINFO_SCU should be something new in devices.h. But I don't see devices.h changes. Maybe it is missed?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 4]
```diff
@@ -1580,6 +1580,64 @@ static int icap_create_subdev_cu(struct platform_device *pdev)
...
+			info.inst_idx = inst++;
+			sprintf(info.iname, "%d",info.inst_idx);
+			info.iname[sizeof(info.iname)-1] = '\0';
+			
+			krnl_info = xocl_query_kernel(xdev, info.kname);
+			if (!krnl_info) {
+				ICAP_WARN(icap, "%s has no metadata. try use default", info.kname);
+				/* Workaround for U30, maybe we can remove this in the future */
+				/*continue;*/
+			}
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:1614]: You don't need xocl_query_kernel() for PS kernel.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 5]
```diff
@@ -1580,6 +1580,64 @@ static int icap_create_subdev_cu(struct platform_device *pdev)
...
+			
+			/* PS kernel do not have base address */
+			info.addr = 0;
+			/* Workaround for U30, maybe we can remove this in the future */
+			info.size = (krnl_info) ? krnl_info->range : 0x1000;
+			if (krnl_info && (krnl_info->features & KRNL_SW_RESET))
+				info.sw_reset = true;
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:1621]: For PS kernel, both address and size should be 0.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 6]
```diff
@@ -1580,6 +1580,64 @@ static int icap_create_subdev_cu(struct platform_device *pdev)
...
+			info.num_res = subdev_info.num_res;
+			info.intr_enable = 0;
+			info.protocol = CTRL_HS;
+			info.intr_id = 0;
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c:1625]: CTRL_HS protocol is PL kernel specific and PS kernel doesn't support interrupt to host yet. But this is not a big deal , let's keep this and make the initial flow work.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/scu.c [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+	info = XOCL_GET_SUBDEV_PRIV(&pdev->dev);
+	BUG_ON(!info);
+	memcpy(&xcu->base.info, info, sizeof(struct xrt_cu_info));
+
+	xcu->base.info.model = XCU_HLS;
+
+	krnl_info = xocl_query_kernel(xdev, info->kname);
+	err = scu_add_args(xcu, krnl_info);
+	if (err)
+		goto err;
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/scu.c:294]: PS kernel doesn't need this.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/scu.c [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+	}
+
+	for (i = 0; i < xcu->base.info.num_res; ++i) {
+		res[i] = platform_get_resource(pdev, IORESOURCE_MEM, i);
+		if (!res[i]) {
+			err = -EINVAL;
+			goto err1;
+		}
+	}
```
Comment (mamin506) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/scu.c:308]: A little bit confusing here. Why PS kernel has IORESOURCE_MEM? I need devices.h change to see if this make sense or not.

**Review by mamin506 — CHANGES_REQUESTED** [review]
A lot of comments. Please merge or rebase to latest master branch then update this PR. Most of the comments are related to it.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/scu.c [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+static void scu_del_args(struct xocl_cu *xcu)
+{
+	if (xcu->base.info.args)
+		vfree(xcu->base.info.args);
+}
+
+static int scu_probe(struct platform_device *pdev)
+{
+	xdev_handle_t xdev = xocl_get_xdev(pdev);
+	struct xocl_cu *xcu;
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/scu.c:266]: Uninitialized variable.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/scu.c [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+
+err:
+	(void) xocl_kds_del_scu(xdev, &xcu->base);
+	return err;
+}
+
+static int scu_remove(struct platform_device *pdev)
+{
+	xdev_handle_t xdev = xocl_get_xdev(pdev);
+	struct xrt_cu_info *info;
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/scu.c:305]: Uninitialized variables.

**Review by rozumx — CHANGES_REQUESTED** [review]
Changes requested.

**Review by uday610 — APPROVED** [review]

**Pull request merged**